fileuploader/large_file_upload_task.go

```diff
@@ -0,0 +1,191 @@
+package fileuploader
+
+import (
+	"context"
+	"errors"
+	abstractions "github.com/microsoft/kiota-abstractions-go"
+	"github.com/microsoft/kiota-abstractions-go/serialization"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+type LargeFileUploadTask[T serialization.Parsable] interface {
+	Upload(progress ProgressCallBack) UploadResult[T]
+	Resume(progress ProgressCallBack) (UploadResult[T], error)
+	Cancel() error
+}
+
+// ByteStream is an interface that represents a stream of bytes
+type ByteStream interface {
+	io.ReaderAt
+	Stat() (os.FileInfo, error)
+}
+
+type largeFileUploadTask[T serialization.Parsable] struct {
+	uploadSession   UploadSession
+	adapter         abstractions.RequestAdapter
+	byteStream      ByteStream // *os.File by default implements ByteStream
+	maxSlice        int64
+	parsableFactory serialization.ParsableFactory
+	errorMappings   abstractions.ErrorMappings
+}
+
+func NewLargeFileUploadTask[T serialization.Parsable](adapter abstractions.RequestAdapter, uploadSession UploadSession, byteStream ByteStream, maxSlice int64, parsableFactory serialization.ParsableFactory, errorMappings abstractions.ErrorMappings) LargeFileUploadTask[T] {
+	return &largeFileUploadTask[T]{
+		adapter:         adapter,
+		uploadSession:   uploadSession,
+		byteStream:      byteStream,
+		maxSlice:        maxSlice,
+		parsableFactory: parsableFactory,
+		errorMappings:   errorMappings,
+	}
+}
+
+// Upload uploads the byteStream in slices and returns the result of the upload
+func (l *largeFileUploadTask[T]) Upload(progress ProgressCallBack) UploadResult[T] {
+	result := NewUploadResult[T]()
+	slices := l.createUploadSlices()
+	maxRetriesPerRequest := 3
+
+	// slices of errors
+	var responseErrors []error
+	var itemResponse T
+
+	var wg sync.WaitGroup
+	wg.Add(len(slices))
+
+	for _, slice := range slices {
+		uploadSlice := slice
+		go func() {
+			defer wg.Done()
+			response, err := l.uploadWithRetry(uploadSlice, maxRetriesPerRequest)
+			if err != nil {
+				responseErrors = append(responseErrors, err)
+			} else {
+				progress(uploadSlice.RangeEnd, uploadSlice.TotalSessionLength)
+			}
+			if response != nil {
+				itemResponse = response.(T)
+			}
+		}()
+	}
+
+	wg.Wait()
+
+	if len(responseErrors) > 0 {
+		result.SetUploadSucceeded(false)
+		result.SetResponseErrors(responseErrors)
+	} else {
+		result.SetUploadSucceeded(true)
```

Comment:
The result property never has the location property set in the event that the API returns a location header with no body like in the outlook api. 
https://learn.microsoft.com/en-us/graph/outlook-large-attachments?tabs=http#response-6